lcls_tools/common/image/fit.py

```diff
@@ -0,0 +1,164 @@
+    min_log_intensity: Optional[float] = 4.0
+    bounding_box_half_width: Optional[PositiveFloat] = 3.0
+    apply_bounding_box_constraint: Optional[bool] = False
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    def fit_image(self, image: ndarray) -> dict:
+        """
+        Public method to determine beam properties from an image, including initial
+        image processing, internal image fitting method, and image validation.
+
+        """
+        image = self.image_processor.auto_process(image)
+        fit_result = self._fit_image(image)
+        fit_result = self._validate_results(image, fit_result)
+        fit_result.update({"processed_image": image})
+        return fit_result
+
+    @abstractmethod
+    def _fit_image(self, image: ndarray) -> dict:
+        """
+        Private image fitting method to be overwritten by subclasses. Expected to
+        return a dictonary containing the following keys:
+        - centroid: 2 element array containing centroid x/y values
+        - rms_sizes: 2 element array containing x/y RMS beamsize values
+        - total_intensity: total pixel intensity
+        """
+        ...
+
+    def _validate_results(self, image: ndarray, fit_results: dict) -> dict:
+        """
+        determine if the image satisfies specified constraints, including:
+        - minimum intensity threshold reached
+        - beam is fully inside the image boundary (if not, RMS measurements are invalid)
+
+        If the beam image does not satisfy these constraints Nan's are returned for
+        beamsize properties and a warning is raised.
+        """
+        if np.log10(fit_results["total_intensity"]) < self.min_log_intensity:
+            # TODO: character count is really not liking this one line
+            warnings.warn(
+                (
+                    "log10 image intensity"
+                    + f"{np.log10(fit_results['total_intensity'])}"
+                    + "below threshold"
+                )
+            )
+            fit_results["centroid"] = np.array((np.Nan, np.Nan))
+            fit_results["rms_sizes"] = np.array((np.Nan, np.Nan))
+            fit_results["bb_penalty"] = np.Nan
+
+        else:
+            centroid = fit_results["centroid"]
+            sizes = fit_results["rms_sizes"]
+
+            if np.all(~np.isnan(np.stack((centroid, sizes)))):
+                # get bounding box penalty
+                bounding_box_penalty = self._calculate_bounding_box_penalty(
+                    image, centroid, sizes,
+                )
+                fit_results["bb_penalty"] = bounding_box_penalty
+
+                # set results to none if the beam extends beyond the roi
+                # and the bounding box constraint is active
+                if bounding_box_penalty > 0 and self.apply_bounding_box_constraint:
+                    fit_results["centroid"] = np.array((np.Nan, np.Nan))
+                    fit_results["rms_sizes"] = np.array((np.Nan, np.Nan))
```

Comment:
Our current practice is to perform scans manually. I could see a value like the bounding box penalty being useful for a human user, but a human user would probably want to preserve the beam parameters, regardless of the quality of the fit. Either we should think more about what the penalty value communicates for a human user, or move the code that parses good and bad data elsewhere.